src/main/java/edu/kit/ipd/crowdcontrol/objectservice/moneytransfer/MailParser.java

```diff
@@ -0,0 +1,101 @@
+package edu.kit.ipd.crowdcontrol.objectservice.moneytransfer;
+
+import edu.kit.ipd.crowdcontrol.objectservice.database.model.tables.records.GiftCodeRecord;
+
+import javax.mail.BodyPart;
+import javax.mail.Message;
+import javax.mail.MessagingException;
+import javax.mail.Multipart;
+import java.io.*;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Describes a parser to extract amazon giftcodes from mails.
+ */
+public class MailParser {
+
+    /**
+     * Extracts a giftcode out of a message.
+     *
+     * @param msg the message, to extract the code
+     * @return returns the giftcode
+     * @throws MoneyTransferException gets thrown, if an error occurred during parsing
+     */
+    protected static Optional<GiftCodeRecord> parseAmazonGiftCode(Message msg) throws MoneyTransferException {
+        //Extract Message
+        String message;
+        Optional<GiftCodeRecord> optional;
+        try {
+            if (!msg.getFrom()[0].toString().toLowerCase().endsWith("amazon.de>")) {
+                return Optional.ofNullable(null);
+            }
+        } catch (MessagingException e) {
+            return Optional.ofNullable(null);
+        }
+        try {
+            Multipart parts = (Multipart) msg.getContent();
+            BodyPart body = parts.getBodyPart(0);
+            Multipart innerMsg = (Multipart) body.getContent();
+            BodyPart textBody = innerMsg.getBodyPart(0);
+            message = textBody.getContent().toString();
+        } catch (ClassCastException | MessagingException | IOException e) {
+            throw new MoneyTransferException("The Parser cannot extract the giftcode from the mails, because the mail format changed. You need to adjust the parser to the new mail format.");
```

Comment:
yes, but i dont get your point. what does the message id give us helpful? And where should i log it? We are already tracking incoming giftcodes and all payments to workers